File: js/20-29/problem23.js
// Finds all abundant numbers up to 'max'
function abundantNumbers(max) {
  var abundant = [];
  for (var i = 1; i <= max; i++) {
     if (Number.divisors(i).sum() > i) {
      abundant.push(i);
    }
  }
  return abundant;
};

// Finds the possible sums of all the numbers in an array up to 'max'.
function possibleSums(arr, max) {
  var sums = [];
  for (var i = 0; i < arr.length; i++) {
    var num1 = arr[i];
    for (var j = i; j < arr.length; j++) {
      if (num1 + arr[j] >= max) {
        continue;
      }
      sums.push(num1 + arr[j]);

    }
  }
  return sums;
};

/* Finds all positive integers which cannot be written
 as the sum of two abundant numbers, up to 'max'. */
function nonAbundantSums(arr, max) {
  var numbers = function() {
    var empty = [];
    for (var i = 1; i < max; i++) {
      empty.push(i);
    }
    return empty;
  }();

  for (var i = 0; i < arr.length; i++) {
    var index = numbers.indexOf(arr[i]);
    if (index >= 0) {
      numbers.splice(index, 1);
    }
  }
  return numbers;
};

console.log(nonAbundantSums(possibleSums(abundantNumbers(28123), 28123), 28123).sum());

File: python/20-29/problem23.py
# Finds all the multiples of a number.
def multiples(n):
    answer = []
    for i in range(2, n):
        if n % i == 0:
            answer.append(i)
            answer.append(n // i)
            if n // i < i:
                break
    if len(answer) > 1:
        answer.pop()
    answer.append(1)
    return set(answer)


# Returns a list of abundant numbers up to max
def abundant_numbers(max):
    return [x for x in range(1, max + 1) if x < sum(multiples(x))]


def non_abundant_sums(n):
    a = abundant_numbers(n)
    sieve = [True] * (n + 1)
    for i in a:
        for j in a:
            k = i + j
            if k > n:
                continue
            sieve[k] = False
    v = [x for x in range(len(sieve)) if sieve[x]]
    return sum(v)


if __name__ == '__main__':
    print(non_abundant_sums(28123))
